### ISSUE TITLE
Error creating an interface proxy in case the interface contains a method with an optional decimal parameter

### ISSUE THREAD

**Issue by fknx** [castleproject/Core#87]
Hello,

I have found an issue in Castle.Core 3.3.3 with the DynamicProxy generation. In case I want to create an interface proxy for an interface with an **optional** decimal parameter, the proxy generation fails with an 'System.Decimal is not a supported constant type' exception. The following program can be used to reproduce the issue:

```csharp
class Program
{
  static void Main(string[] args)
  {
	 var proxyGenerator = new ProxyGenerator();
	 proxyGenerator.CreateInterfaceProxyWithoutTarget<IMyClass>(new StandardInterceptor());
  }
}

public interface IMyClass
{
  void Do(decimal d = 0m);
}
```

However, the exception only occurs with the Castle.Core build targeting the **.NET Framework 4.5**. The .NET Framework 4.0 build works fine. (I have not tested the 3.5 or Silverlight versions).

Here you can find the full stacktrace:
```
System.Reflection.Emit.TypeBuilder.SetConstantValue(ModuleBuilder module, Int32 tk, Type destType, Object value)
Castle.DynamicProxy.Generators.Emitters.MethodEmitter.DefineParameters(ParameterInfo[] parameters)
Castle.DynamicProxy.Generators.Emitters.MethodEmitter..ctor(AbstractTypeEmitter owner, String name, MethodAttributes attributes, MethodInfo methodToUseAsATemplate)
Castle.DynamicProxy.Generators.Emitters.AbstractTypeEmitter.CreateMethod(String name, MethodAttributes attributes, MethodInfo methodToUseAsATemplate)
Castle.DynamicProxy.Generators.MethodGenerator.Generate(ClassEmitter class, ProxyGenerationOptions options, INamingScope namingScope)
Castle.DynamicProxy.Contributors.CompositeTypeContributor.ImplementMethod(MetaMethod method, ClassEmitter class, ProxyGenerationOptions options, OverrideMethodDelegate overrideMethod)
Castle.DynamicProxy.Contributors.CompositeTypeContributor.Generate(ClassEmitter class, ProxyGenerationOptions options)
Castle.DynamicProxy.Generators.InterfaceProxyWithoutTargetGenerator.GenerateType(String typeName, Type proxyTargetType, Type[] interfaces, INamingScope namingScope)
Castle.DynamicProxy.Generators.InterfaceProxyWithTargetGenerator.<>c__DisplayClass1.<GenerateCode>b__0(String n, INamingScope s)
Castle.DynamicProxy.Generators.BaseProxyGenerator.ObtainProxyType(CacheKey cacheKey, Func`3 factory)
Castle.DynamicProxy.Generators.InterfaceProxyWithTargetGenerator.GenerateCode(Type proxyTargetType, Type[] interfaces, ProxyGenerationOptions options)
Castle.DynamicProxy.DefaultProxyBuilder.CreateInterfaceProxyTypeWithoutTarget(Type interfaceToProxy, Type[] additionalInterfacesToProxy, ProxyGenerationOptions options)
Castle.DynamicProxy.ProxyGenerator.CreateInterfaceProxyTypeWithoutTarget(Type interfaceToProxy, Type[] additionalInterfacesToProxy, ProxyGenerationOptions options)
Castle.DynamicProxy.ProxyGenerator.CreateInterfaceProxyWithoutTarget(Type interfaceToProxy, Type[] additionalInterfacesToProxy, ProxyGenerationOptions options, IInterceptor[] interceptors)
Castle.DynamicProxy.ProxyGenerator.CreateInterfaceProxyWithoutTarget(Type interfaceToProxy, IInterceptor interceptor)
Castle.DynamicProxy.ProxyGenerator.CreateInterfaceProxyWithoutTarget[TInterface](IInterceptor interceptor)
ConsoleApplication.Program.Main(String[] args) in Program.cs:Zeile 12.
System.AppDomain._nExecuteAssembly(RuntimeAssembly assembly, String[] args)
Microsoft.VisualStudio.HostingProcess.HostProc.RunUsersAssembly()
System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
System.Threading.ThreadHelper.ThreadStart()
```

Best regards,
Frederik

**Comment by jonorossi** [comment]
The exception only occurs in the .NET 4.5 build because the code that sets the default value for optional parameters is conditionally compiled out in prior framework builds because the `SetConstant` method wouldn't have been available:
* https://github.com/castleproject/Core/blob/master/src/Castle.Core/DynamicProxy/Generators/Emitters/MethodEmitter.cs#L159-L164

This looks like a bug in the .NET Framework 4.5 because the `TypeBuilder` doesn't handle `decimal`, and I'm not even sure which `CorElementType` would be reported for decimals:
* http://referencesource.microsoft.com/#mscorlib/system/reflection/emit/typebuilder.cs,485
* https://msdn.microsoft.com/en-us/library/ms232600(v=vs.110).aspx

**Comment by fknx** [comment]
I agree that it is pretty weird that a decimal may not be passed to the `SetConstant` method, but I am not sure that it is a bug. At least it is documented on [MSDN](https://msdn.microsoft.com/de-de/library/system.reflection.emit.parameterbuilder.setconstant%28v=vs.110%29.aspx):

> defaultValue is restricted to the following types: Boolean, SByte, Int16, Int32, Int64, Byte, UInt16, UInt32, UInt64, Single, Double, DateTime, Char, String, and Enum.

I am not sure about the consequences as I am not familiar with the `TypeBuilder`, but would it be possible to simply omit the `SetConstant` call for all decimals?

**Comment by jonorossi** [comment]
Ah yer, that method would have been there since early .NET days because the CLR has always supported optional parameters for COM interop.

I suspect that the documentation was written from the code. If you look at the doco for .NET Framework 2.0 it actually explicitly calls out Decimal should be null, however I think that has been fixed in later revisions of the documentation because [passing null would actually throw an exception](http://referencesource.microsoft.com/#mscorlib/system/reflection/emit/typebuilder.cs,524).

> defaultValue is restricted to the following types: Boolean, SByte, Int16, Int32, Int64, Byte, UInt16, UInt32, UInt64, Single, Double, DateTime, Char, String, and Enum. **If the type of the parameter is Decimal or Object, defaultValue can only be null**.

I see no reason why DP shouldn't just omit calls to `SetConstant` for decimals with a big glaring comment referring to the fact that the framework code can't handle them so it'll have to do. We work around a big pile of defects right through DP's code, so another one is no problem.

**Comment by fknx** [comment]
Sorry for responding this late, but I have now commited a [fix](https://github.com/fknx/Core/commit/3fb21eca970e022dfb97c5385238872f452d4168) for this issue and issued a pull request. Right now the issue does not affect us (as we are stuck with .NET 4.0) but it would be great if the fix could be added to a future release :)

Thank you for your help!

**Comment by jonorossi** [comment]
Fixed by #91. Thanks.
